api/plugins/app_group_lifecycle.py

```diff
@@ -0,0 +1,529 @@
+            members: The list of users that were removed from the group.
+            plugin_id: If provided, only the plugin matching this ID should respond.
+                       If None, all plugins may respond.
+        """
+
+    @hookspec
+    def sync_all_group_membership(self, session: Session, app: App, plugin_id: str | None = None) -> None:
+        """
+        Bulk sync all group memberships for an app. This is invoked periodically by a CLI command `flask sync-app-group-memberships`.
+
+        Args:
+            session: The Access database session.
+            app: The app for which to sync all group membership.
+            plugin_id: If provided, only the plugin matching this ID should respond.
+                       If None, all plugins may respond.
+        """
+
+
+def get_app_group_lifecycle_hook() -> pluggy.HookRelay:
+    """Get the hook relay for app group lifecycle plugins."""
+    global _cached_app_group_lifecycle_hook
+
+    if _cached_app_group_lifecycle_hook is not None:
+        return _cached_app_group_lifecycle_hook
+
+    pm = pluggy.PluginManager(app_group_lifecycle_plugin_name)
+    pm.add_hookspecs(AppGroupLifecyclePluginSpec)
+
+    # Register the hook wrappers
+    pm.register(sys.modules[__name__])
+
+    count = pm.load_setuptools_entrypoints(app_group_lifecycle_plugin_name)
+    logger.info(f"Loaded {count} app group lifecycle plugin(s)")
+
+    _cached_app_group_lifecycle_hook = pm.hook
+
+    return _cached_app_group_lifecycle_hook
+
+
+def get_app_group_lifecycle_plugins() -> list[AppGroupLifecyclePluginMetadata]:
+    """
+    Get a registry of all loaded app group lifecycle plugins with their metadata.
+
+    Returns:
+        A list of plugin metadata objects.
+    """
+    global _cached_plugin_registry
+
+    if _cached_plugin_registry is not None:
+        return _cached_plugin_registry
+
+    hook = get_app_group_lifecycle_hook()
+
+    # Collect metadata from all registered plugins
+    plugins: list[AppGroupLifecyclePluginMetadata] = [
+        plugin for plugin in hook.get_plugin_metadata() if plugin is not None
+    ]
+
+    # Validate uniqueness
+    seen_ids: set[str] = set()
+    seen_display_names: set[str] = set()
+    seen_descriptions: set[str] = set()
+
+    for plugin in plugins:
+        if not plugin.id:
+            raise ValueError("Plugin ID is required")
+        if not plugin.display_name:
+            raise ValueError(f"Display name is required but missing for plugin {plugin.id}")
+        if not plugin.description:
+            raise ValueError(f"Description is required but missing for plugin {plugin.id}")
+
+        if plugin.id in seen_ids:
+            raise ValueError(f"Duplicate plugin ID detected: {plugin.id}")
+        if plugin.display_name in seen_display_names:
+            raise ValueError(f"Duplicate plugin display name detected: {plugin.display_name}")
+        if plugin.description in seen_descriptions:
```

Comment:
We're probably not going to see collisions of plugin descriptions unless someone copy-pastas one as a template, we probably okay not requiring uniqueness, but we can leave this as this is minor either way.